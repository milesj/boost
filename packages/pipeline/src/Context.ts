import { isObject, isPlainObject } from '@boost/common';

interface Cloneable {
  clone?: () => unknown;
}

export default class Context {
  /**
   * Create a new instance of the current context and shallow clone all properties.
   */
  clone(...args: any[]): this {
    // @ts-expect-error
    const context = new this.constructor(...args);

    // Copy enumerable properties
    Object.keys(this).forEach((key) => {
      const prop = key as keyof this;
      let value: unknown = this[prop];

      if (Array.isArray(value)) {
        value = [...value];
      } else if (value instanceof Map) {
        value = new Map(value);
      } else if (value instanceof Set) {
        value = new Set(value);
      } else if (value instanceof Date) {
        value = new Date(value.getTime());
      } else if (isObject<Cloneable>(value)) {
        if (typeof value.clone === 'function') {
          value = value.clone();
          // Dont dereference instances, only plain objects
        } else if (isPlainObject(value)) {
          value = { ...value };
        }
      }

      context[prop] = value;
    });

    return context;
  }
}
