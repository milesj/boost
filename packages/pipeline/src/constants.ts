import { createInternalDebugger } from '@boost/internal';
import { Status } from './types';

export const debug = createInternalDebugger('pipeline');

export const STATUS_PENDING: Status = 'pending';
export const STATUS_RUNNING: Status = 'running';
export const STATUS_SKIPPED: Status = 'skipped';
export const STATUS_PASSED: Status = 'passed';
export const STATUS_FAILED: Status = 'failed';
