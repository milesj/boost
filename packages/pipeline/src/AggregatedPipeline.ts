import ParallelPipeline from './ParallelPipeline';
import Context from './Context';
import { AggregatedResult } from './types';

export default class AggregatedPipeline<
  Ctx extends Context,
  Input,
  Output = Input
> extends ParallelPipeline<{}, Ctx, Input, Output> {
  blueprint() {
    return {};
  }

  /**
   * Execute all work units in parallel with a value being passed to each work unit.
   * Work units will synchronize regardless of race conditions and errors.
   */
  async run(): Promise<AggregatedResult<Output>> {
    const { context, value } = this;
    const work = this.getWorkUnits();

    this.debug('Aggregating %d work units', work.length);
    this.onRun.emit([value]);

    return Promise.all(
      work.map(unit => {
        this.onRunWorkUnit.emit([unit, value]);

        return unit.run(context, value).catch(error => error);
      }),
    ).then(responses => this.aggregateResult(responses));
  }
}
