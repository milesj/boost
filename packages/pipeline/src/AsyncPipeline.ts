import { Contract, instanceOf } from '@boost/common';
import Context from './Context';
import { Action, AggregatedResult, Runnable } from './types';
import createWorkUnit from './createWorkUnit';

export default abstract class AsyncPipeline<
  Options extends object,
  Input,
  Output = Input,
  Ctx extends Context = Context
> extends Contract<Options> {
  queue: Runnable<Input, Output>[] = [];

  value: Input;

  constructor(value: Input, options?: Options) {
    super(options);

    this.value = value;
  }

  /**
   * Enqueue a work unit to the end of the queue.
   */
  push(title: string, action: Action<Ctx, Input, Output>): this;
  push(workUnit: Runnable<Input, Output>): this;
  push(
    titleOrWorkUnit: string | Runnable<Input, Output>,
    action?: Action<Ctx, Input, Output>,
  ): this {
    this.queue.push(createWorkUnit(titleOrWorkUnit, action));

    return this;
  }

  /**
   * Aggregate and partition errors and results into separate collections.
   */
  protected aggregateResult(responses: (Error | Output)[]): AggregatedResult<Output> {
    const errors: Error[] = [];
    const results: Output[] = [];

    responses.forEach(response => {
      if (instanceOf(response, Error)) {
        errors.push(response);
      } else {
        results.push(response);
      }
    });

    return { errors, results };
  }

  /**
   * Run and process the entire work unit queue.
   */
  abstract async run(context: Ctx): Promise<any>;
}
