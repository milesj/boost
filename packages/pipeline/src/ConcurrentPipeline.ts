import ParallelPipeline from './ParallelPipeline';
import Context from './Context';
import debug from './debug';

export default class ConcurrentPipeline<
  Ctx extends Context,
  Input = unknown,
  Output = Input
> extends ParallelPipeline<{}, Ctx, Input, Output> {
  blueprint() {
    return {};
  }

  /**
   * Execute all work units in parallel with a value being passed to each work unit.
   * If an error occurs, the pipeline will abort early, otherwise return a list of all results.
   */
  async run(): Promise<Output[]> {
    const { context, value } = this;
    const work = this.getWorkUnits();

    debug('Running %d in parallel', work.length);

    this.onRun.emit([value]);

    const result = await Promise.all(
      work.map((unit) => {
        this.onRunWorkUnit.emit([unit, value]);

        return unit.run(context, value);
      }),
    );

    this.onFinish.emit([]);

    return result;
  }
}
