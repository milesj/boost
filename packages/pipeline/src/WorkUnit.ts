import { Contract } from '@boost/common';
import { BailEvent, Event } from '@boost/event';
import {
  STATUS_FAILED,
  STATUS_PASSED,
  STATUS_PENDING,
  STATUS_RUNNING,
  STATUS_SKIPPED,
} from './constants';
import { Context } from './Context';
import { PipelineError } from './PipelineError';
import { Action, Hierarchical, Runnable, Status } from './types';

export abstract class WorkUnit<Options extends object, Input = unknown, Output = Input>
  extends Contract<Options>
  implements Runnable<Input, Output>, Hierarchical
{
  depth: number = 0;

  index: number = 0;

  output?: Output;

  input?: Input;

  startTime: number = 0;

  statusText: string = '';

  stopTime: number = 0;

  /**
   * Called when an execution fails.
   * @category Events
   */
  readonly onFail = new Event<[Error | null, Input]>('fail');

  /**
   * Called when an execution succeeds.
   * @category Events
   */
  readonly onPass = new Event<[Output, Input]>('pass');

  /**
   * Called before a work unit is executed. Can return `true` to skip the work unit.
   * @category Events
   */
  readonly onRun = new BailEvent<[Input]>('run');

  /**
   * Called when an execution is skipped.
   * @category Events
   */
  readonly onSkip = new Event<[Input]>('skip');

  readonly title: string;

  private action: Action<Context, Input, Output>;

  private status: Status = STATUS_PENDING;

  // We want to support all contexts, so we use any.
  // Unknown and `Context` will not work because of the constraint.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  constructor(title: string, action: Action<any, Input, Output>, options?: Options) {
    super(options);

    if (!title || typeof title !== 'string') {
      throw new PipelineError('WORK_REQUIRED_TITLE');
    }

    if (action !== null && typeof action !== 'function') {
      throw new PipelineError('ACTION_REQUIRED');
    }

    this.action = action;
    this.status = STATUS_PENDING;
    this.title = title;
  }

  /**
   * Return true if the task failed when executing.
   */
  hasFailed(): boolean {
    return this.status === STATUS_FAILED;
  }

  /**
   * Return true if the task executed successfully.
   */
  hasPassed(): boolean {
    return this.status === STATUS_PASSED;
  }

  /**
   * Return a unique hierarchical ID.
   */
  get id() {
    return `work[${this.depth}:${this.index}]`;
  }

  /**
   * Return true if the task has been completed in any form.
   */
  isComplete(): boolean {
    return this.hasPassed() || this.hasFailed() || this.isSkipped();
  }

  /**
   * Return true if the task has not been executed yet.
   */
  isPending(): boolean {
    return this.status === STATUS_PENDING;
  }

  /**
   * Return true if the task is currently running.
   */
  isRunning(): boolean {
    return this.status === STATUS_RUNNING;
  }

  /**
   * Return true if the task was or will be skipped.
   */
  isSkipped(): boolean {
    return this.status === STATUS_SKIPPED;
  }

  /**
   * Run the current task by executing it and performing any before and after processes.
   */
  async run(context: Context, value: Input): Promise<Output> {
    this.input = value;
    const skip = this.onRun.emit([value]);
    const runner: Action<Context, Input, Output> = this.action;

    if (skip || this.isSkipped() || !runner) {
      this.status = STATUS_SKIPPED;
      this.onSkip.emit([value]);

      // Allow input as output. This is problematic for skipping
      // since the expected output is no longer in sync. Revisit.
      // @ts-expect-error Allow invalid type
      return Promise.resolve(value);
    }

    this.status = STATUS_RUNNING;
    this.startTime = Date.now();

    try {
      this.output = await runner(context, value, this);
      this.status = STATUS_PASSED;
      this.stopTime = Date.now();
      this.onPass.emit([this.output, value]);
    } catch (error: unknown) {
      this.status = STATUS_FAILED;
      this.stopTime = Date.now();

      if (error instanceof Error) {
        this.onFail.emit([error, value]);

        throw error;
      }
    }

    this.statusText = '';

    return this.output!;
  }

  /**
   * Mark a task as skipped if the condition is true.
   */
  skip(condition: boolean = true): this {
    if (condition) {
      this.status = STATUS_SKIPPED;
    }

    return this;
  }
}
