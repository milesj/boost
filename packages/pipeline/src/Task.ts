import { Blueprint, Predicates } from '@boost/common';
import WorkUnit from './WorkUnit';

export default class Task<Input = unknown, Output = Input> extends WorkUnit<{}, Input, Output> {
	// A task is simply a work unit that executes a standard function.
	// It doesn't need configurable options, so implement an empty blueprint.
	blueprint(preds: Predicates): Blueprint<object> {
		return {};
	}
}
