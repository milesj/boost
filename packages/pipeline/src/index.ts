/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import AsyncPipeline from './AsyncPipeline';
import ConcurrentPipeline from './ConcurrentPipeline';
import Context from './Context';
import PooledPipeline, { PooledOptions } from './PooledPipeline';
import Routine from './Routine';
import SyncPipeline from './SyncPipeline';
import SynchronizedPipeline from './SynchronizedPipeline';
import Task from './Task';
import WaterfallPipeline from './WaterfallPipeline';
import WorkUnit from './WorkUnit';

export * from './constants';
export * from './types';

export {
  AsyncPipeline,
  ConcurrentPipeline,
  Context,
  PooledPipeline,
  PooledOptions,
  Routine,
  SyncPipeline,
  SynchronizedPipeline,
  Task,
  WaterfallPipeline,
  WorkUnit,
};
