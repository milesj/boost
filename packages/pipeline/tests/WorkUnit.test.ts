import Context from '../src/Context';
import Task from '../src/Task';
import WorkUnit from '../src/WorkUnit';
import { STATUS_RUNNING } from '../src/constants';

describe('Work', () => {
  let context: Context;
  let passWork: WorkUnit<{}, number, number>;
  let failWork: WorkUnit<{}, string, string>;

  class CountContext extends Context {
    count = 0;
  }

  beforeEach(() => {
    context = new Context();
    passWork = new Task('title', (ctx, value) => value * 2);
    failWork = new Task<string, string>('title', () => {
      throw new Error('Oops');
    });
  });

  describe('constructor()', () => {
    it('errors if no title', () => {
      expect(() => new Task('', () => {})).toThrowErrorMatchingSnapshot();
    });

    it('errors if title is not a string', () => {
      expect(
        () =>
          new Task(
            // @ts-expect-error
            123,
            () => {},
          ),
      ).toThrowErrorMatchingSnapshot();
    });

    it('errors if action is not a function', () => {
      expect(
        () =>
          new Task(
            'title',
            // @ts-expect-error
            123,
          ),
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe('hasFailed()', () => {
    it('returns a boolean for STATUS_FAILED status state', async () => {
      expect(failWork.hasFailed()).toBe(false);

      try {
        await failWork.run(context, '');
      } catch {
        // Ignore
      }

      expect(failWork.hasFailed()).toBe(true);
    });
  });

  describe('hasPassed()', () => {
    it('returns a boolean for STATUS_PASSED status state', async () => {
      expect(passWork.hasPassed()).toBe(false);

      await passWork.run(context, 0);

      expect(passWork.hasPassed()).toBe(true);
    });
  });

  describe('id()', () => {
    it('returns hierarchy ID', () => {
      passWork.depth = 8;
      passWork.index = 3;

      expect(passWork.id).toBe('work[8:3]');
    });
  });

  describe('isComplete()', () => {
    it('returns true when passed', async () => {
      expect(passWork.isComplete()).toBe(false);

      await passWork.run(context, 0);

      expect(passWork.isComplete()).toBe(true);
    });

    it('returns true when failed', async () => {
      expect(failWork.isComplete()).toBe(false);

      try {
        await failWork.run(context, '');
      } catch {
        // Ignore
      }

      expect(failWork.isComplete()).toBe(true);
    });

    it('returns true when skipped', async () => {
      expect(passWork.isComplete()).toBe(false);

      await passWork.skip().run(context, 0);

      expect(passWork.isComplete()).toBe(true);
    });

    it('returns false when pending (default)', () => {
      expect(passWork.isComplete()).toBe(false);
    });

    it('returns false when running', () => {
      expect(passWork.isComplete()).toBe(false);

      // @ts-expect-error
      passWork.status = STATUS_RUNNING;

      expect(passWork.isComplete()).toBe(false);
    });
  });

  describe('isPending()', () => {
    it('returns a boolean for STATUS_PENDING status state (default)', () => {
      expect(passWork.isPending()).toBe(true);
    });
  });

  describe('isRunning()', () => {
    it('returns a boolean for STATUS_RUNNING status state', () => {
      expect(passWork.isRunning()).toBe(false);

      // @ts-expect-error
      passWork.status = STATUS_RUNNING;

      expect(passWork.isRunning()).toBe(true);
    });
  });

  describe('isSkipped()', () => {
    it('returns a boolean for STATUS_SKIPPED status state', () => {
      expect(passWork.isSkipped()).toBe(false);

      passWork.skip();

      expect(passWork.isSkipped()).toBe(true);
    });
  });

  describe('run()', () => {
    it('stores action result as output', async () => {
      const value = await passWork.run(context, 123);

      expect(value).toBe(246);
      expect(passWork.output).toBe(value);
    });

    it('resolves a value with the action', async () => {
      try {
        expect(await passWork.run(context, 123)).toBe(246);
      } catch (error) {
        expect(true).toBe(false);
      }
    });

    it('resolves a value if the task should be skipped', async () => {
      try {
        passWork.skip();

        expect(await passWork.run(context, 123)).toBe(123);
      } catch (error) {
        expect(true).toBe(false);
      }
    });

    it('rejects the value if the action throws an error', async () => {
      try {
        await failWork.run(context, '');

        expect(true).toBe(false);
      } catch (error) {
        expect(error).toEqual(new Error('Oops'));
        expect(failWork.hasFailed()).toBe(true);
      }
    });

    it('passes a context to the action', async () => {
      context = new CountContext();
      passWork = new Task('title', (ctx: CountContext) => {
        ctx.count = 1;

        return ctx.count;
      });

      await passWork.run(context, 123);

      expect(context).toEqual(
        expect.objectContaining({
          count: 1,
        }),
      );
    });

    it('sets times on success', async () => {
      await passWork.run(context, 123);

      expect(passWork.startTime).not.toBe(0);
      expect(passWork.stopTime).not.toBe(0);
    });

    it('sets times on failure', async () => {
      try {
        await failWork.run(context, '');
      } catch (error) {
        // Ignore
      }

      expect(failWork.startTime).not.toBe(0);
      expect(failWork.stopTime).not.toBe(0);
    });

    it('passes work unit as 3rd argument to action', async () => {
      const spy = jest.fn();

      passWork = new Task('title', spy);

      await passWork.run(context, 123);

      expect(spy).toHaveBeenCalledWith({}, 123, passWork);
    });

    it('emits `onSkip` event when skipped', async () => {
      const spy = jest.fn();

      passWork.onSkip.listen(spy);

      await passWork.skip().run(context, 123);

      expect(spy).toHaveBeenCalledWith(123);
    });

    it('emits `onRun` event before running', async () => {
      const spy = jest.fn();

      passWork.onRun.listen(spy);

      await passWork.run(context, 123);

      expect(spy).toHaveBeenCalledWith(123);
    });

    it('emits `onPass` event on success', async () => {
      const spy = jest.fn();

      passWork.onPass.listen(spy);

      await passWork.run(context, 123);

      expect(spy).toHaveBeenCalledWith(246);
    });

    it('emits `onFail` event on error', async () => {
      const spy = jest.fn();

      failWork.onFail.listen(spy);

      try {
        await failWork.run(context, '');
      } catch (error) {
        expect(spy).toHaveBeenCalledWith(error);
      }
    });

    it('can skip if `onRun` listener returns false', async () => {
      const skipSpy = jest.fn();
      const otherSpy = jest.fn();

      passWork.onRun.listen(() => false);
      passWork.onSkip.listen(skipSpy);
      passWork.onPass.listen(otherSpy);
      passWork.onFail.listen(otherSpy);

      await passWork.run(context, 123);

      expect(passWork.isSkipped()).toBe(true);
      expect(skipSpy).toHaveBeenCalledWith(123);
      expect(otherSpy).not.toHaveBeenCalled();
    });
  });

  describe('skip()', () => {
    it('marks a task as skipped', () => {
      expect(passWork.isSkipped()).toBe(false);

      passWork.skip();

      expect(passWork.isSkipped()).toBe(true);
    });

    it('evaluates a condition to determine whether to skip', () => {
      expect(passWork.isSkipped()).toBe(false);

      // @ts-expect-error
      passWork.skip(1 === 2);

      expect(passWork.isSkipped()).toBe(false);
    });
  });
});
