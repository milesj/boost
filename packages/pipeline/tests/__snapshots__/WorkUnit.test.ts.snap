// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Work constructor() errors if action is not a function 1`] = `"Work units require an executable function. [PLN:ACTION_REQUIRED]"`;

exports[`Work constructor() errors if no title 1`] = `"Work units require a title. [PLN:WORK_REQUIRED_TITLE]"`;

exports[`Work constructor() errors if title is not a string 1`] = `"Work units require a title. [PLN:WORK_REQUIRED_TITLE]"`;
